pkg/hashreader/hashreader.go

```diff
@@ -0,0 +1,120 @@
+package hashreader
+
+// HashReader writes what it reads from an io.Reader to an MD5 and SHA256 hash.Hash.
+import (
+	"bytes"
+	"crypto/md5"
+	"crypto/sha256"
+	"encoding/hex"
+	"hash"
+	"io"
+)
+
+// HashReader verifies that the content of the io.Reader matches the expected checksums.
+type HashReader struct {
+	md5sum, sha256sum []byte // Byte values of md5sum, sha256sum of client sent values.
+
+	src                 io.Reader
+	size                int64
+	md5Hash, sha256Hash hash.Hash
+}
+
+// New returns a new HashReader computing the MD5 sum and SHA256 sum
+// (if set) of the provided io.Reader.
+func New(src io.Reader, size int64, md5Hex, sha256Hex string) (*HashReader, error) {
+	sha256sum, err := hex.DecodeString(sha256Hex)
+	if err != nil {
+		return nil, err
+	}
+	md5sum, err := hex.DecodeString(md5Hex)
+	if err != nil {
+		return nil, err
+	}
+	if hr, ok := src.(*HashReader); ok {
+		return hr, nil
+	}
+	var (
+		md5Hash    hash.Hash
+		sha256Hash hash.Hash
+	)
+	if len(md5sum) != 0 {
+		md5Hash = md5.New()
+	}
+	if len(sha256sum) != 0 {
+		sha256Hash = sha256.New()
+	}
+	return &HashReader{
+		md5sum:     md5sum,
+		sha256sum:  sha256sum,
+		src:        io.LimitReader(src, size),
+		size:       size,
+		md5Hash:    md5Hash,
+		sha256Hash: sha256Hash,
+	}, nil
+}
+
+func (r *HashReader) Read(p []byte) (n int, err error) {
+	n, err = r.src.Read(p)
+	if err != nil && err != io.EOF {
+		return
+	}
+
+	if r.md5Hash != nil {
+		r.md5Hash.Write(p[:n])
+	}
+	if r.sha256Hash != nil {
+		r.sha256Hash.Write(p[:n])
+	}
+
+	// At io.EOF verify if the checksums are right.
+	if err == io.EOF {
+		cerr := r.Verify()
+		if cerr != nil {
+			return 0, cerr
+		}
+		return n, err
+	}
+
+	return
+}
+
+// Size returns the absolute number of bytes the HashReader
+// will return during reading. It returns -1 for unlimited
+// data.
+func (r *HashReader) Size() int64 { return r.size }
+
+// MD5 - returns byte md5 value
+func (r *HashReader) MD5() []byte {
+	return r.md5sum
+}
+
+// SHA256 - returns byte sha256 value
+func (r *HashReader) SHA256() []byte {
+	return r.sha256sum
+}
+
+// MD5Hex returns hex md5 value.
+func (r *HashReader) MD5Hex() string {
+	return hex.EncodeToString(r.md5sum)
+}
+
+// SHA256Hex returns hex sha256 value.
+func (r *HashReader) SHA256Hex() string {
+	return hex.EncodeToString(r.sha256sum)
+}
+
+// Verify verifies if the computed MD5 sum and SHA256 sum are
+// equal to the ones specified when creating the HashReader.
+func (r *HashReader) Verify() error {
```

Comment:
Do we need this method as public?  There is no meaning someone calls in the middle

Reply:
This is still needed there is a bug in erasure code which is not fixed yet.